src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/ZLibStreamPool.cs

```diff
@@ -0,0 +1,283 @@
+            try
+            {
+                if (!ExecutionContext.IsFlowSuppressed())
+                {
+                    ExecutionContext.SuppressFlow();
+                    restoreFlow = true;
+                }
+
+                // There is no need to use weak references here, because these pools are kept
+                // for the entire lifetime of the application. Also we reset the timer on each tick,
+                // which prevents the object being rooted forever.
+                _cleaningTimer = new Timer(x => ((ZLibStreamPool)x!).RemoveStaleItems(),
+                    state: this, Timeout.Infinite, Timeout.Infinite);
+            }
+            finally
+            {
+                if (restoreFlow)
+                {
+                    ExecutionContext.RestoreFlow();
+                }
+            }
+        }
+
+        public static ZLibStreamPool GetOrCreate(int windowBits)
+        {
+            Debug.Assert(windowBits >= WebSocketDeflateOptions.MinWindowBits
+                      && windowBits <= WebSocketDeflateOptions.MaxWindowBits);
+
+            int index = windowBits - WebSocketDeflateOptions.MinWindowBits;
+            ref ZLibStreamPool? pool = ref s_pools[index];
+
+            return Volatile.Read(ref pool) ?? EnsureInitialized(windowBits, ref pool);
+
+            static ZLibStreamPool EnsureInitialized(int windowBits, ref ZLibStreamPool? target)
+            {
+                Interlocked.CompareExchange(ref target, new ZLibStreamPool(windowBits), null);
+
+                Debug.Assert(target != null);
+                return target;
+            }
+        }
+
+        public ZLibStreamHandle GetInflater()
+        {
+            if (TryGet(_inflaters, out ZLibStreamHandle? stream))
+            {
+                return stream;
+            }
+
+            return CreateInflater();
+        }
+
+        public void ReturnInflater(ZLibStreamHandle stream)
+        {
+            if (stream.InflateReset() != ErrorCode.Ok)
+            {
+                stream.Dispose();
+                return;
+            }
+
+            Return(stream, _inflaters);
+        }
+
+        public ZLibStreamHandle GetDeflater()
+        {
+            if (TryGet(_deflaters, out ZLibStreamHandle? stream))
+            {
+                return stream;
+            }
+
+            return CreateDeflater();
+        }
+
+        public void ReturnDeflater(ZLibStreamHandle stream)
+        {
+            if (stream.DeflateReset() != ErrorCode.Ok)
+            {
+                stream.Dispose();
+                return;
+            }
+
+            Return(stream, _deflaters);
+        }
+
+        private void Return(ZLibStreamHandle stream, List<CacheItem> cache)
+        {
+            lock (cache)
+            {
+                if (cache.Count < MaximumRetained)
+                {
+                    cache.Add(new CacheItem(stream));
+
+                    if (Interlocked.Increment(ref _activeCount) == 1)
```

Comment:
What will be the difference between `_activeCount` and `cache.Count`?

Reply:
_activeCount are inflaters + deflaters. It's only used to control whether the cleaning timer should be active or not. Also it is used with interlocked, because at the moment the pool allows for concurrent use for inflaters and deflaters - i.e. renting a deflater will not block renting an inflater.